working/4271 - static enough metaprogramming/proposal.md

```diff
@@ -0,0 +1,1222 @@
+    for (@konst final field in TypeInfo.of<T>().fields)
+      if (!field.isStatic) field.name: field.getFrom(value),
+  };
+```
+
+```dart
+// Example
+class A {
+  final int a;
+  final String b;
+  A({required this.a, required this.b});
+}
+
+// Calling toJson<A>(A(...)) produces specialization
+Map<...> toJson$A(A value) => {a: value.a, b: value.b};
+```
+
+#### `fromJson<@konst T>`
+
+```dart
+T fromJson<@konst T>(Map<String, Object?> json) {
+  final typeInfo = TypeInfo.of<T>();
+  return invoke<T>(
+        typeMirror.defaultConstructor,
+        [],
+        named: {
+          for (@konst final field in typeInfo.fields)
+            if (!field.isStatic)
+              field.name: invoke(
+                _valueFromJson,
+                [json[field.name]],
+                types: [field.type.type],
+              ),
+        },
+      );
+}
+
+FieldType _valueFromJson<@konst FieldType>(Object? value) {
+  var fieldType = TypeInfo.of<FieldType>();
+  if (fieldType.isNullable) {
+    if (value == null) {
+      return null as FieldType;
+    }
+    fieldType = fieldType.nonNullable;
+  } else {
+    if (value == null) {
+      throw ArgumentError('Field not found in incoming json');
+    }
+  }
+
+  // Primitive values are mapped directly.
+  if (fieldType.isSubtypeOf<String>() ||
+      fieldType.isSubtypeOf<num>() ||
+      fieldType.isSubtypeOf<bool>()) {
+    return value as FieldType;
+  }
+
+  // Lists are unpacked element by element.
+  if (fieldType.instantiationOf<List>() case final instantiation?) {
+    final elementType = instantiation.typeArguments.first.type;
+    return invoke<FieldType>(
+          listFromJson,
+          [value as List<Object?>],
+          typeArguments: [elementType],
+        );
+  } else {
+    // We assume that this is Map -> class conversion then.
+    return fromJson<FieldType>(value as Map<String, Object?>);
+  }
+}
+
+List<E> _listFromJson<@konst E>(List<Object?> list) {
+  return <E>[for (var v in list) _valueFromJson<E>(v)];
+}
+```
+
+```dart
+// Example
+class A {
+  final int a;
+  final String b;
+  A({required this.a, required this.b});
+}
+
+// Calling fromJson<A>({...}) produces specializations:
+
+A fromJson$A(Map<String, Object?> map) {
+  return A(
+    a: _valueFromJson$int(map['a']),
+    b: _valueFromJson$String(map['a']),
+  );
+}
+
+int _valueFromJson$int(Object? value) {
+  if (value == null) {
+    throw ArgumentError('Field not found in incoming json');
+  }
+  return value as int;
+}
+
+String _valueFromJson$String(Object? value) {
+  if (value == null) {
+    throw ArgumentError('Field not found in incoming json');
+  }
+  return value as String;
+}
+```
+
+### Example: Defining `hashCode` and `==`
+
+We could also instruct compiler to handle `mixin`'s (and possibly all generic
+classes) with `@konst` type parameters in a special way: _clone_ their
+declarations with known type arguments. This would allow to write the following
+code:
+
+```dart
+mixin DataClass<@konst T> {
```

Comment:
This would be 🔥

Reply:
Yeah, it would be useful in many places. Also it would solve a performance problem we occasionally encounter where inheritance or generics cause polymorphism and severely reduce performance.